libkbfs/block_retrieval_queue.go

```diff
@@ -221,47 +225,62 @@ func (brq *blockRetrievalQueue) CacheAndPrefetch(ctx context.Context,
 		dbc := brq.config.DiskBlockCache()
 		if dbc != nil {
 			go func() {
-				err := dbc.UpdateMetadata(ctx, ptr.ID, hasPrefetched)
+				err := dbc.UpdateMetadata(ctx, ptr.ID, hasPrefetched, donePrefetch)
+				close(didUpdateCh)
 				switch err.(type) {
 				case nil:
 				case NoSuchBlockError:
 				default:
 					brq.log.CWarningf(ctx, "Error updating metadata: %+v", err)
 				}
 			}()
+		} else {
+			close(didUpdateCh)
 		}
 	}()
 	if hasPrefetched {
 		return brq.config.BlockCache().PutWithPrefetch(ptr, kmd.TlfID(), block,
 			lifetime, hasPrefetched)
 	}
-	if priority < defaultOnDemandRequestPriority {
-		// Only on-demand or higher priority requests can trigger prefetches.
+	if priority < lowestTriggerPrefetchPriority {
+		// Only high priority requests can trigger prefetches.
 		hasPrefetched = false
 		return brq.config.BlockCache().PutWithPrefetch(ptr, kmd.TlfID(), block,
 			lifetime, hasPrefetched)
 	}
-	// We must let the cache know at this point that we've prefetched.
-	// 1) To prevent any other Gets from prefetching.
-	// 2) To prevent prefetching if a cache Put fails, since prefetching if
-	//	  only useful when combined with the cache.
+	// To prevent any other Gets from prefetching, we must let the cache know
+	// at this point that we've prefetched.
 	hasPrefetched = true
+	donePrefetch = false
 	err = brq.config.BlockCache().PutWithPrefetch(ptr, kmd.TlfID(), block,
 		lifetime, hasPrefetched)
 	switch err.(type) {
 	case nil:
 	case cachePutCacheFullError:
-		brq.log.CDebugf(ctx, "Skipping prefetch because the cache "+
-			"is full")
-		return err
+		brq.log.CWarningf(ctx, "Cache is full, but prefetching anyway to "+
+			"populate the disk cache.")
```

Comment:
That fix was from before the disk cache. We'll get errors when trying to place the blocks into the in-memory cache, but at least we can put them in the disk cache so that we don't have to re-retrieve them later.